sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_outlier_filter/cuda_polar_voxel_outlier_filter.cu

```diff
@@ -0,0 +1,949 @@
+  {
+    return opt.has_value() ? opt.value() : ::cuda::std::numeric_limits<DataType>::max();
+  }
+};
+
+struct NulloptToLowest
+{
+  /*
+   * Functor to think cuda::std::nullopt as ::cuda::std::numeric_limits<int>::lowest()
+   * NOTE: std::numeric_limits::min returns the minimum, non-zero, positive value while
+   * std::numeric_limits::lowest() returns the negative minimum value
+   */
+  template <typename DataType>
+  __host__ __device__ __forceinline__ DataType
+  operator()(const ::cuda::std::optional<DataType> & opt) const
+  {
+    return opt.has_value() ? opt.value() : ::cuda::std::numeric_limits<DataType>::lowest();
+  }
+};
+
+/** \brief factory utility function to allocate unique_ptr and FieldDataComposer with allocated
+ * pointer
+ */
+template <typename T>
+[[nodiscard]] std::tuple<
+  CudaPooledUniquePtr<T>, CudaPooledUniquePtr<T>, CudaPooledUniquePtr<T>, FieldDataComposer<T *>>
+generate_field_data_composer(const size_t & num_elems, cudaStream_t & stream, cudaMemPool_t & pool)
+{
+  auto radius = autoware::cuda_utils::make_unique<T>(num_elems, stream, pool);
+  auto azimuth = autoware::cuda_utils::make_unique<T>(num_elems, stream, pool);
+  auto elevation = autoware::cuda_utils::make_unique<T>(num_elems, stream, pool);
+
+  FieldDataComposer<T *> composer;
+  composer.radius = radius.get();
+  composer.azimuth = azimuth.get();
+  composer.elevation = elevation.get();
+  return std::make_tuple(
+    std::move(radius), std::move(azimuth), std::move(elevation), std::move(composer));
+}
+
+/** \brief atomicAdd operation for size_t
+ *
+ * Since cuda's builtin atomicAdd does not have overloaded function for size_t,
+ * and the definition may differ according to platforms, this function define atomicAdd operation
+ * with size check
+ */
+__device__ void atomic_add_size_t(size_t * addr, size_t val)
+{
+  if constexpr (sizeof(size_t) == sizeof(unsigned int)) {
+    atomicAdd(reinterpret_cast<unsigned int *>(addr), static_cast<unsigned int>(val));
+  } else if constexpr (sizeof(size_t) == sizeof(unsigned long long int)) {
+    atomicAdd(
+      reinterpret_cast<unsigned long long int *>(addr), static_cast<unsigned long long int>(val));
+  } else {
+    static_assert(
+      sizeof(size_t) == sizeof(unsigned int) || sizeof(size_t) == sizeof(unsigned long long int),
+      "atomicAdd_size_t is only supported for size_t sizes equal to unsigned int or unsigned long "
+      "long int.");
+  }
+}
+
+__device__ [[nodiscard]] inline bool meets_primary_threshold(
+  const size_t & count, const int & threshold)
+{
+  return count >= static_cast<size_t>(threshold);
+}
+
+__device__ [[nodiscard]] inline bool meets_secondary_threshold(
+  const size_t & count, const int & threshold)
+{
+  return count <= static_cast<size_t>(threshold);
+}
+
+__device__ [[nodiscard]] inline bool meets_intensity_threshold(
+  const uint8_t intensity, const uint8_t threshold)
+{
+  return intensity <= threshold;
+}
+
```

Comment:
The naming is a bit confusing for me. As far as I understand, secondary returns below the threshold get filtered out. So, I would expect that anything **above** the threshold is considered to "meet the threshold". This naming suggests the opposite. Maybe something like `is_below_threshold` would be clearer?